### ISSUE TITLE
False positive for S4158, collection is known to be empty

### ISSUE THREAD

**Issue by gilles-leblanc** [SonarSource/sonar-csharp#1042]
### Description

I'm using the SonarLint Visual Studio extension and I am getting a false positive on some C# code. For S4158 "Remove this call, the collection is known to be empty". 

I have tested my intuition that this was a false positive by putting a breakpoint and debugging and Visual Studio and I was right, the breakpoint got hit regardless of what SonarLint said.

While I am using SonarLint for Visual Studio, the related repo's issues seems to be more related to the software itself and not the lint rules. I decided it was best to submit the issue to this repo (sonar-csharp) rather than the repo for the Visual Studio Extension.

### Repro steps

I can't divulge the code used to get this false positive but I will try to reproduce some other code as similar as possible to the original.

```
// code to get businessObject from database, if null create a new businessObject
// ...

List<FirstEntity> firstList = new List<FirstEntity>();

foreach (var x in businessObject.SubCollection)
{
    var secondEntity = (from y in context.SecondEntity
                                  where y.Foo = "Bar"
                                  select y).FirstOrDefault();
    if (secondEntity == null)
        firstList.Add(x);
}

firstList.ForEach(x => context.FirstEntity.Remove(x));    // the false positive is on this line
```

### Expected behavior

This false positive should not have been flagged as an error.

### Actual behavior

The false positive is flagged as an error.

### Known workarounds

Ignore the false positive.

### Related information

* SonarC# Version : Using SonarLint for Visual Studio 2015 version 3.7.0.2645
* Visual Studio Version : Professional 2015, Version 14.0.25424.00 Update 3

**Comment by valhristov** [comment]
Thank you for your feedback, indeed this is the right place to submit issues about the rules.

This is a known limitation of the rule and in our dataflow engine in general - our support for loops is poor and the engine cannot detect that you are actually adding items to `firstList` inside the `foreach` loop, hence it raises the false positive. Improving loop support is high on our backlog for this year and we will be definitely working on it in the upcoming months, but for now the only option is to ignore the FP or disable the rule (because frankly speaking and even though I wrote it, it does not bring much value in its current state).

**Comment by valhristov** [comment]
Duplicate of #767
